core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -0,0 +1,234 @@
+             samDAO: SamDAO,
+             googleStorage: GoogleStorageService[IO],
+             pubSubDao: GooglePubSubDAO,
+             pubSubSubscriptionName: String,
+             avroUpsertBucketName: String,
+             batchSize: Int)(implicit cs: ContextShift[IO]): Props =
+    Props(new AvroUpsertMonitorActor(pollInterval, pollIntervalJitter, workspaceService, googleServicesDAO, samDAO, googleStorage, pubSubDao, pubSubSubscriptionName, avroUpsertBucketName, batchSize))
+}
+
+class AvroUpsertMonitorActor(
+                                  val pollInterval: FiniteDuration,
+                                  pollIntervalJitter: FiniteDuration,
+                                  workspaceService: UserInfo => WorkspaceService,
+                                  googleServicesDAO: GoogleServicesDAO,
+                                  samDAO: SamDAO,
+                                  googleStorage: GoogleStorageService[IO],
+                                  pubSubDao: GooglePubSubDAO,
+                                  pubSubSubscriptionName: String,
+                                  avroUpsertBucketName: String,
+                                  batchSize: Int)(implicit cs: ContextShift[IO])
+  extends Actor
+    with LazyLogging
+    with FutureSupport {
+  import AvroUpsertMonitor._
+  import context._
+
+  case class AvroMetadataJson(namespace: String, name: String, userSubjectId: String, userEmail: String, jobId: String)
+  implicit val avroMetadataJsonFormat = jsonFormat5(AvroMetadataJson)
+
+  self ! StartMonitorPass
+
+  // fail safe in case this actor is idle too long but not too fast (1 second lower limit)
+  setReceiveTimeout(max((pollInterval + pollIntervalJitter) * 10, 1 second))
+
+  private def max(durations: FiniteDuration*): FiniteDuration = {
+    implicit val finiteDurationIsOrdered = scala.concurrent.duration.FiniteDuration.FiniteDurationIsOrdered
+    durations.max
+  }
+
+  override def receive = {
+    case StartMonitorPass =>
+      // start the process by pulling a message and sending it back to self
+      pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+
+    case Some(message: PubSubMessage) =>
+      //we received a message, so we will parse it and try to upsert it
+      logger.info(s"received avro upsert message: $message")
+      val (jobId, file) = parseMessage(message)
+
+      file match {
+        case "upsert.json" => initAvroUpsert(jobId, message.ackId).map(_ => ImportComplete) pipeTo self
+        case _ => acknowledgeMessage(message.ackId).map(_ => None) pipeTo self //some other file (i.e. success/failure log, metadata.json)
+      }
+
+    case None =>
+      // there was no message so wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime, self, StartMonitorPass)
+
+    case ImportComplete => self ! None
```

Comment:
We just completed an import which may have taken a few minutes. Let's `StartMonitorPass` right away. `None` will wait a poll interval before checking the topic again.